realm/realm-library/src/main/java/io/realm/internal/StatefulCollectionChangeSet.java

```diff
@@ -55,52 +50,37 @@ public State getState() {
 
     @Override
     public int[] getInsertions() {
-        if (state == State.INITIAL || state == State.INITIAL_INCOMPLETE) {
-            return NO_INDEX_CHANGES;
-        } else {
-            return changeset.getInsertions();
-        }
+        return changeset.getInsertions();
     }
 
     @Override
     public int[] getChanges() {
-        if (state == State.INITIAL || state == State.INITIAL_INCOMPLETE) {
-            return NO_INDEX_CHANGES;
-        } else {
-            return changeset.getChanges();
-        }
+        return changeset.getChanges();
     }
 
     @Override
     public Range[] getDeletionRanges() {
-        if (state == State.INITIAL || state == State.INITIAL_INCOMPLETE) {
-            return NO_RANGE_CHANGES;
-        } else {
-            return changeset.getDeletionRanges();
-        }
+        return changeset.getDeletionRanges();
     }
 
     @Override
     public Range[] getInsertionRanges() {
-        if (state == State.INITIAL || state == State.INITIAL_INCOMPLETE) {
-            return NO_RANGE_CHANGES;
-        } else {
-            return changeset.getInsertionRanges();
-        }
+        return changeset.getInsertionRanges();
     }
 
     @Override
     public Range[] getChangeRanges() {
-        if (state == State.INITIAL || state == State.INITIAL_INCOMPLETE) {
-            return NO_RANGE_CHANGES;
-        } else {
-            return changeset.getChangeRanges();
-        }
+        return changeset.getChangeRanges();
     }
 
     @Nullable
     @Override
     public Throwable getError() {
         return error;
     }
+
+    @Override
+    public boolean isCompleteResult() {
+        return remoteDataSynchronized;
```

Comment:
Why is this idempotent? shouldn't we call `changeset.isRemoteDataLoaded()` again to check if the data download finished?

Reply:
Each changeset is a value object, so if this value changes a new ChangeSet object is created.